Parallel attention with flexbert modules

NohTow opened the pull request:
Implementation of #2 using Flexbert modules.

iacolippo commented:
LGTM

iacolippo commented on src/bert_layers/attention.py:
```diff
@@ -333,6 +333,112 @@ def forward(
+        self.all_head_size = self.num_attention_heads * self.attn_head_size
+        self.p_dropout = config.attention_probs_dropout_prob
+        # Compute QKV and FF outputs at once
+        self.Wqkvff = nn.Linear(config.hidden_size, 3 * self.all_head_size + config.intermediate_size, bias=config.attn_qkv_bias)
```
FWIW I don't think we will ever use bias, so I would remove it entirely for the sake of simplicity

warner-benjamin replied:
All the linear layers currently have a bias option. I'd leave it in for consistency.

warner-benjamin commented on src/bert_layers/layers.py:
```diff
@@ -292,6 +292,32 @@ def forward(
+            hidden_states: (total_nnz, dim)
+            attn_mask: None or (batch, max_seqlen)
+        """
+        attn_out, intermediate_ff = self.attn(self.norm(hidden_states), cu_seqlens, max_seqlen, indices, attn_mask)
+        return hidden_states + attn_out + self.mlp(intermediate_ff)
```
I think this implementation won't dispatch the attention and mlp in parallel because the mlp layer might be blocked until the attention calculation is finished, since both `attn_out` and `intermediate_ff` are required before the mlplayer can start processing `intermediate_ff`.

warner-benjamin reviewed (changes requested):
I'd move the fused linear layer into the ParallelPreNorm layer and remove the potential sequential dependency between the attention and mlp layers so they can dispatch in parallel.

Something like this:

```python
class FlexBertUnpadParallelPreNormLayer(FlexBertLayerBase):
    """Composes the FlexBERT parallel attention and MLP blocks into a single layer using pre-normalization."""

    def __init__(self, config: FlexBertConfig):
        super().__init__()
        self.attn_size = config.hidden_size * 3
        self.mlp_size = config.intermediate_size
        self.norm = get_norm_layer(config)
        self.Wqkvff = nn.Linear(config.hidden_size, self.attn_size + self.mlp_size, bias=config.attn_qkv_bias)
        self.attn = get_attention_layer(config)
        self.mlp = get_mlp_layer(config)

    def forward(
        self,
        hidden_states: torch.Tensor,
        cu_seqlens: torch.Tensor,
        max_seqlen: int,
        indices: Optional[torch.Tensor] = None,
        attn_mask: Optional[torch.Tensor] = None,
    ) -> torch.Tensor:
        attn_in, mlp_in = self.Wqkvff(self.norm(hidden_states)).split([self.attn_size, self.mlp_size], dim=1)
        return hidden_states + self.attn(attn_in, cu_seqlens, max_seqlen, indices, attn_mask) + self.mlp(mlp_in)
```

NohTow commented:
I changed the layout to prevent any blocking computation (I find it cleaner also anyways), rebased on main (with a little oopsie), fixed the attention in my layers and added the padded version.

warner-benjamin reviewed (approved):
LGTM
